Fix: disorder and FEAT: arc_thickness_by (DFG)

pablogw opened the pull request:
Description

  This PR fixes the diagram disorder issue in Multi-Perspective DFG visualizations and adds a new feature to control arc
  thickness.

  Changes:
  - Fix: Corrects event sorting to ensure stable ordering when timestamps are identical, preventing diagram disorder
  - Feature: Adds arc_thickness_by parameter to control arc thickness based on "frequency", "time", or "none"
  - Updates dfg_builder.py to sort by both start_timestamp and timestamp in stable order
  - Extends GraphVizDiagrammer with methods to calculate arc thickness based on the selected perspective

  How to Test

  1. Load an event log with events that have identical timestamps
  2. Generate a Multi-Perspective DFG using mpvis.mpdfg.discover_multi_perspective_dfg()
  3. Visualize with mpvis.mpdfg.view_multi_perspective_dfg() or mpvis.mpdfg.save_vis_multi_perspective_dfg()
  4. Test the new arc_thickness_by parameter with values: "frequency", "time", "none"
  5. Verify that diagrams are consistently ordered (no disorder between runs)
  6. Verify that arc thickness reflects the chosen perspective correctly

  Additional Information

  - Related issues: Fixes diagram disorder bug
  - Modified files:
    - src/mpvis/mpdfg/dfg_builder.py: Stable sorting fix
    - src/mpvis/mpdfg/diagrammers/graphviz.py: Arc thickness control
    - src/mpvis/mpdfg/actions.py: New parameter API
  - Breaking changes: None (backward compatible - defaults to "frequency")

pablogw committed:
Fix: disorder and FEAT: arc_thickness_by (DFG)

nicoabarca commented on src/mpvis/mpdfg/diagrammers/graphviz.py:
```diff
@@ -27,6 +27,7 @@ def __init__(
         visualize_cost: bool = True,
         cost_currency: str = "USD",
         rankdir: str = "TB",
+        arc_thickness_by: str = "frequency",
```
[REQ]
Use string literals.

```suggestion
    arc_thickness_by: Literal["frequency", "time"] = "frequency",
```

nicoabarca commented on src/mpvis/mpdfg/diagrammers/graphviz.py:
```diff
@@ -207,5 +198,35 @@ def build_activity_name(self, activity_name: str) -> str:
 
         return activity_name
 
+    def get_arc_thickness_for_connection(self, connection):
```
[REQ]

When instantiating `GraphVizDiagrammer` you define `self.arc_thickness_by` to `"frequency"` or `"time"`, so there is no need to check what the parameter value is. The function logic could be a lot cleaner doing something like:

```suggestion
    def get_arc_thickness_for_connection(self, connection):
        try:
            return link_width(
                    self.dfg["connections"][connection][self.arc_thickness_by],
                    self.connections_dimensions_min_and_max[self.arc_thickness_by],
                    )   
        except KeyError: # if by any chance self.arc_thickness_by value is not a key in self.dfg["connections"][connection] or self.connections_dimensions_min_and_max
            return 1
```

nicoabarca commented on src/mpvis/mpdfg/diagrammers/graphviz.py:
```diff
@@ -207,5 +198,35 @@ def build_activity_name(self, activity_name: str) -> str:
+        else:  # "none" or any other value
+            return 1
+    
+    def get_arc_thickness_for_extreme(self, frequency):
```
[REQ]
Make function logic cleaner. Something like this can work:

```suggestion
    def get_arc_thickness_for_extreme(self, frequency):
        if self.arc_thickness_by == "frecuency" and isinstance(frequency, (int, float)):
            return link_width(frequency, self.connections_dimensions_min_and_max["frequency"])
        return 1 # for every other case return value will be 1
```
I think there is no need to check "frequency" key in self.connections_dimensions_min_and_max.

nicoabarca commented on src/mpvis/mpdfg/actions.py:
```diff
@@ -138,6 +138,7 @@ def get_multi_perspective_dfg_string(
     cost_currency: str = "USD",
     rankdir: str = "TD",
     diagram_tool: str = "graphviz",
+    arc_thickness_by: str = "frequency",
```
[REQ]
Use string literals.

```suggestion
    arc_thickness_by: Literal["frequency", "time"] = "frequency",
```

nicoabarca commented on src/mpvis/mpdfg/actions.py:
```diff
@@ -199,6 +202,7 @@ def view_multi_perspective_dfg(
     cost_currency: str = "USD",
     rankdir: str = "TD",
     format: str = "svg",
+    arc_thickness_by: str = "frequency",
```
[REQ]
Use string literals.

```suggestion
    arc_thickness_by: Literal["frequency", "time"] = "frequency",
```

nicoabarca commented on src/mpvis/mpdfg/actions.py:
```diff
@@ -250,6 +256,7 @@ def save_vis_multi_perspective_dfg(
     format: str = "svg",
     rankdir: str = "TD",
     diagram_tool: str = "graphviz",
+    arc_thickness_by: str = "frequency",
```
[REQ]
Use string literals.

```suggestion
    arc_thickness_by: Literal["frequency", "time"] = "frequency",
```

nicoabarca commented on src/mpvis/mpdfg/dfg_builder.py:
```diff
@@ -13,8 +13,11 @@ def __init__(self, dfg, log, parameters):
+        sorting_order.append(self.parameters.start_timestamp_key)
+        sorting_order.append(self.parameters.timestamp_key)
+        # Asegurar orden estable para casos con timestamps idénticos
+        sorted_log = self.log.sort_values(by=sorting_order, kind='stable') 
```
[NIT]
Use double quotes for strings.

nicoabarca replied:
[QUESTION]
Does this approach correctly order events at the "middle" of a case execution? I remember that we only checked an event log where we have problems with start activities.

nicoabarca replied:
[QUESTION]
Does this approach correctly order events at the "middle" of a case execution? I remember that we only checked an event log where we have problems with start activities.

nicoabarca commented on src/mpvis/mpdfg/dfg_builder.py:
```diff
@@ -13,8 +13,11 @@ def __init__(self, dfg, log, parameters):
     def start(self):
-        sorted_log = self.log.sort_values(by=[self.parameters.timestamp_key])
-        sorted_log = self.log.sort_values(by=[self.parameters.start_timestamp_key])
+        sorting_order = []
```
[REQ]
No need to construct list dynamically with appends. Just do:

```suggestion
        sorting_order = [self.parameters.start_timestamp_key, self.parameters.timestamp_key]
```

nicoabarca commented on main.py:
```diff
@@ -1,184 +1,195 @@
 import pandas as pd
```
[REQ]
Don't push changes to this file. It only exists for testing purposes.

[SUG]
If you want, delete this file.

nicoabarca commented on uv.lock:
```diff
@@ -1,5 +1,5 @@
 version = 1
```
[REQ]
Revert changes to this file. This file should only change when updating/removing dependencies.

nicoabarca reviewed (commented):
Good job! Code looks great. Only some small request changes.

nicoabarca reviewed (changes requested):
Address above changes and ask me for a second review.

pablogw committed:
Fix: fix comments pr3

pablogw committed:
Remove unused code and comments from main.py

pablogw committed:
Revert changes

nicoabarca commented on src/mpvis/mpdfg/dfg_builder.py:
```diff
@@ -13,8 +13,11 @@ def __init__(self, dfg, log, parameters):
+        sorting_order = [self.parameters.start_timestamp_key, self.parameters.timestamp_key]
+        sorting_order.append(self.parameters.start_timestamp_key)
+        sorting_order.append(self.parameters.timestamp_key)
+        # Asegurar orden estable para casos con timestamps idénticos
```
[REQ]
Remove comment in Spanish

nicoabarca commented on src/mpvis/mpdfg/diagrammers/graphviz.py:
```diff
@@ -207,5 +199,20 @@ def build_activity_name(self, activity_name: str) -> str:
+    def get_arc_thickness_for_extreme(self, frequency):
+        if self.arc_thickness_by == "frecuency" and isinstance(frequency, (int, float)):
+            return link_width(frequency, self.connections_dimensions_min_and_max["frequency"])
+        return 1 # for every other case return value will be 1
```
[REQ]
Remove comment

nicoabarca commented on src/mpvis/mpdfg/dfg_builder.py:
```diff
@@ -13,8 +13,11 @@ def __init__(self, dfg, log, parameters):
-        sorted_log = self.log.sort_values(by=[self.parameters.start_timestamp_key])
+        sorting_order = [self.parameters.start_timestamp_key, self.parameters.timestamp_key]
+        sorting_order.append(self.parameters.start_timestamp_key)
+        sorting_order.append(self.parameters.timestamp_key)
```
[REQ]
Remove appends. We already defined the sorting order with the corresponding elements.

pablogw committed:
fix: address PR #3 review feedback

- Remove redundant sorting_order appends in dfg_builder.py
- Remove Spanish comment
- Fix typo: frecuency -> frequency in graphviz.py
- Simplify exception handling comments
- Add comprehensive test suite for event ordering with identical timestamps

Tests verify stable sorting works correctly at START, MIDDLE, and END of case execution.

nicoabarca committed:
feat: ordering tests and test dependencies

nicoabarca committed:
chore: formatting

nicoabarca commented:
LGTM 🚀
